packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts

```diff
@@ -0,0 +1,277 @@
+  /**
+   * The Amazon Resource Name (ARN) of the target.
+   */
+  readonly arn: string;
+
+  /**
+   * Role to use to invoke this event target
+   */
+  readonly role: iam.IRole;
+
+  /**
+   *  What input to pass to the tatget
+   */
+  readonly input?: ScheduleTargetInput;
+
+  /**
+   * A `RetryPolicy` object that includes information about the retry policy settings, including the maximum age of an event, and the maximum number of times EventBridge Scheduler will try to deliver the event to a target.
+   */
+  readonly retryPolicy?: CfnSchedule.RetryPolicyProperty;
+
+  /**
+   * An object that contains information about an Amazon SQS queue that EventBridge Scheduler uses as a dead-letter queue for your schedule. If specified, EventBridge Scheduler delivers failed events that could not be successfully delivered to a target to the queue.\
+   */
+  readonly deadLetterConfig?: CfnSchedule.DeadLetterConfigProperty
+
+  /**
+   *  The templated target type for the Amazon ECS RunTask API Operation.
+   */
+  readonly ecsParameters?: CfnSchedule.EcsParametersProperty;
+  /**
+   * The templated target type for the EventBridge PutEvents API operation.
+   */
+  readonly eventBridgeParameters?: CfnSchedule.EventBridgeParametersProperty;
+
+  /**
+   * The templated target type for the Amazon Kinesis PutRecord API operation.
+   */
+  readonly kinesisParameters?: CfnSchedule.KinesisParametersProperty;
+
+  /**
+   * The templated target type for the Amazon SageMaker StartPipelineExecution API operation.
+   */
+  readonly sageMakerPipelineParameters?: CfnSchedule.SageMakerPipelineParametersProperty;
+  /**
+   * The templated target type for the Amazon SQS SendMessage API Operation
+   */
+  readonly sqsParameters?: CfnSchedule.SqsParametersProperty;
+}
+
+/**
+ * Bind props to base schedule target config.
+ * @internal
+ */
+export function bindBaseTargetConfig(props: ScheduleTargetBaseProps) {
+  let { deadLetterQueue } = props;
+
+  return {
+    deadLetterConfig: deadLetterQueue ? { arn: deadLetterQueue?.queueArn } : undefined,
+    retryPolicy: renderRetryPolicy(props.maxEventAge, props.retryAttempts),
+  };
+}
+
+const roleCache: {[key: string]: iam.Role} = {};
+
+/**
+ * Obtain the Role for the EventBridge Scheduler event
+ *
+ * If a role already exists, it will be returned. This ensures that if multiple
+ * events have the same target, they will share a role.
+ * @internal
+ */
+function singletonScheduleRole(schedule: ISchedule, targetArn: string): iam.IRole {
+  const stack = Stack.of(schedule);
+  const arn = Token.isUnresolved(targetArn) ? stack.resolve(targetArn).toString() : targetArn;
+  const hash = md5hash(arn).slice(0, 6);
+  const id = 'SchedulerRoleForTarget-' + hash;
+  const existing = stack.node.tryFindChild(id) as iam.IRole;
+
+  const principal = new iam.PrincipalWithConditions(new iam.ServicePrincipal('scheduler.amazonaws.com'), {
+    StringEquals: {
+      'aws:SourceAccount': schedule.env.account,
```

Comment:
@Jacco I have removed condition for checking sourceArn as it created a circular dependency between schedule -> target -> role -> schedule. 

In your original prototype it was implemented as: 
```
 const scheduleArn = Arn.format({
    service: 'scheduler',
    resourceName: `${schedule.group.groupName}/${schedule.scheduleName}`,
    resource: 'schedule'
  }, Stack.of(schedule));
  const principal = new iam.PrincipalWithConditions(new iam.ServicePrincipal('scheduler.amazonaws.com'), {
    'StringEquals': {
        "aws:SourceArn": scheduleArn,
        "aws:SourceAccount": schedule.env.account,
    }
  })
```

It didn't create a dependency cycle in your code, as now the `scheduleName` is a reference to the CFN resource instead of just a string:
```
    this.scheduleName = this.getResourceNameAttribute(resource.ref);
```